soco/zonegroupstate.py

```diff
@@ -148,8 +150,84 @@ def poll(self, soco):
+        except SoCoUPnPException as soco_upnp_exception:
+            _LOG.debug(
+                "Exception (%s) raised on 'GetZoneGroupState()'",
+                soco_upnp_exception,
+            )
+
+            if config.ZGT_EVENT_FALLBACK is False:
+                _LOG.debug("ZGT event fallback disabled (config.ZGT_EVENT_FALLBACK)")
+                raise NotSupportedException(
+                    "'GetZoneGroupState()' call fails on large Sonos systems "
+                    "and event fallback is disabled"
+                ) from soco_upnp_exception
+
+            _LOG.debug("Falling back to using a ZGT event")
+            try:
+                self.update_zgs_by_event(soco)
+            except Exception as soco_exception:
+                raise soco_exception from soco_upnp_exception
+
+    def update_zgs_by_event(self, speaker):
+        """
+        Fall back to updating the ZGS using a ZGT event.
+        Use of the 'events_twisted' module is not currently supported.
+        """
+        if config.EVENTS_MODULE.__name__ == "soco.events":
+            _LOG.debug("Updating ZGS using standard 'events' module")
+            self.update_zgs_by_event_default(speaker)
+
+        if config.EVENTS_MODULE.__name__ == "soco.events_asyncio":
+            _LOG.debug("Updating ZGS using 'events_asyncio' module")
+            # Explicit asyncio event loop control required for Python 3.6
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            loop.run_until_complete(ZoneGroupState.update_zgs_by_event_asyncio(speaker))
+            asyncio.set_event_loop(None)
+            loop.close()
+            # From Python 3.7, we can just use the single statement:
+            # asyncio.run(ZoneGroupState.update_zgs_events_asyncio(speaker))
+
+        if config.EVENTS_MODULE.__name__ == "soco.events_twisted":
+            raise SoCoException(
+                "ZGT event fallback not supported when using the "
+                "'events_twisted' module"
+            )
+
+    def update_zgs_by_event_default(self, speaker):
+        """
+        Update the ZGS using the default events module.
+        """
+        sub = speaker.zoneGroupTopology.subscribe()
+        event = sub.events.get(timeout=1.0)
+        sub.unsubscribe()
+        zgs = event.variables.get("zone_group_state")
+        self.process_payload(payload=zgs, source="event", source_ip=speaker.ip_address)
+
+    @staticmethod
+    async def update_zgs_by_event_asyncio(speaker):
+        """
+        Update ZGS using events_asyncio. When the event is received,
+        the events_asyncio library will call 'process_payload' with
+        the updated ZGS.
+        """
+        from . import events_asyncio  # pylint: disable=C0415
+
+        sub = await speaker.zoneGroupTopology.subscribe()
+        await asyncio.sleep(1.0)
+        await sub.unsubscribe()
+        await events_asyncio.event_listener.async_stop()
```

Comment:
What happens here if other subscriptions were created elsewhere?

Reply:
Good point. The `async_stop()` will stop the event listener for all subscriptions. I'll add a test to stop the event listener only if it was not running prior to the `subscribe()`, in which case it would have been started by our `subscribe()` call.